source/client/output_collector_impl.cc

```diff
@@ -122,5 +122,51 @@ std::string YamlOutputCollectorImpl::toString() const {
   return Envoy::MessageUtil::getYamlStringFromMessage(toProto(), true, true);
 }
 
+DottedStringOutputCollectorImpl::DottedStringOutputCollectorImpl(Envoy::TimeSource& time_source,
+                                                                 const Options& options)
+    : OutputCollectorImpl(time_source, options) {}
+
+std::string DottedStringOutputCollectorImpl::toString() const {
+  std::stringstream ss;
+  const auto& output = toProto();
+  for (const auto& result : output.results()) {
+    for (const auto& statistic : result.statistics()) {
+      const std::string prefix = fmt::format("{}.{}", result.name(), statistic.id());
+
+      ss << fmt::format("{}.samples: {}", prefix, statistic.count()) << std::endl;
+      ss << fmt::format("{}.mean: {}", prefix,
+                        Envoy::Protobuf::util::TimeUtil::DurationToMicroseconds(statistic.mean()))
+         << std::endl;
+      ss << fmt::format("{}.pstdev: {}", prefix,
+                        Envoy::Protobuf::util::TimeUtil::DurationToMicroseconds(statistic.pstdev()))
+         << std::endl;
+
+      // The proto percentiles are ordered ascending. We write the first match to the stream.
+      // Revisit to see if we can factor out a common denominator with the CLI output formatter.
+      double last_percentile = -1.;
+      for (const double p : {.5, .75, .8, .9, .95, .99, .999, 1.}) {
```

Comment:
This iteration over the histogram seems like a common pattern shared by the different output formatters. Would it be possible to create a higher-order function or class that implements this, with the inner block supplied by a lambda?

Reply:
Sure; can I take a TODO on that?

Reply:
I think it could be good to do in this PR, it should actually shrink the size of the PR.

Reply:
Done in https://github.com/envoyproxy/nighthawk/pull/150/commits/3836ba1cc1e39f0265e7ec02714a6de417a4fb30